Update0.2.3

shawndove opened the pull request.

wlangera reviewed (commented):
Parsing dataframe does not work yet.
I got stuck at the part where renaming of column names occurs.

At line 277 columns are renamed assuming no extra columns in the input dataframe.
E.g. gcube has a column geometry. 
Should we allow extra columns or not? It would be nice to keep extra columns to give the user a bit more flexibility.
We could put them after all essential columns.

Also, you assume that the order of the column names in the input dataframe has the same order as your `col_names_defaultlist`. So this does not work.

shawndove commented on R/process_cube_functions.R:
```diff
@@ -101,13 +106,27 @@ process_cube <- function(cube_name,
-  )
+  data_type = match.arg(data_type)
+
+  if (data_type == "csv") {
```
Good idea. I have taken your advice here, except using base R to avoid adding another dependency.

shawndove commented on R/process_cube_functions.R:
```diff
@@ -164,6 +183,12 @@ process_cube <- function(cube_name,
+  } else if (grid_type == "none") {
+
+      # create dummy column full of zeros
+      occurrence_data$cellCode <- 0
```
I thought the issue was that there are no grid codes? In any case, I got rid of this code so you can attach your own cell codes column..

I fixed the issue with column name order. The problem was actually a coding error that mixed the column names into the wrong order rather than an assumption of any particular order.

Extra columns are already allowed, there should not be an issue there..?

shawndove commented:
Sorry it took so long, but I think I've addressed the issues now. I tried making a basic cube using gcube and was able to run it through process_cube without errors. Please let me know if you still have trouble.

wlangera commented:
Hi

Thank you. 
I think we are almost there!

Please install the latest gcube version 0.3.0 (`remotes::install_github("b-cubed-eu/gcube")`) for the following code to work.
You can use the simulated cube from this code for debugging.
Issues:

1. Date Range is not processed correctly. I have 6 time points (1, 2, 3, 4, 5, 6) but the metadata says 1-5 and the calculation of a time series indicator only gives values for 1-5.
2. I would change the argument `grid_type = "none"` because this seems like there is not grid reference system. Yet, you can make cubes using gcube with any grid reference system you want. Should we rename it to `grid_type = "custom"`? Or something else more clear?
3.  Number of cells: **--> not calculated?** yet, `cols_cellCode = "id"` is provided
4. Number of species represented: 6 **--> ok**
    Number of families represented:  **--> not calculated?** yet, `cols_family = "family"` is provided
    Kingdoms represented: **--> not calculated?** yet, `cols_kingdom = "kingdom"` is provided

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(sf)
#> Linking to GEOS 3.12.1, GDAL 3.8.4, PROJ 9.3.1; sf_use_s2() is TRUE
library(gcube)
library(b3gbi)

# Create a polygon to simulate occurrences
polygon <- st_polygon(list(cbind(c(500, 1000, 1000, 600, 200, 100, 500),
                                 c(200, 100, 700, 1000, 900, 500, 200))))

# Create grid for grid designation
cube_grid <- st_make_grid(
  st_buffer(polygon, 50),
  n = c(20, 20),
  square = TRUE) %>%
  st_sf()

# Create dataframe with simulation function arguments
multi_species_dataset <- tibble(
  plgn = rep(list(polygon), 6),
  n_time_points = rep(6, 6),
  detection_probability = rep(c(0.8, 0.9, 1), 2),
  coords_uncertainty_meters = rep(c(25, 30, 50), 2),
  grid = rep(list(cube_grid), 6),
  seed = 123
)

# Add taxonomic hierarchy and generate cube
map_occ_cube_df <- multi_species_dataset %>%
  generate_taxonomy(num_genera = 4, num_families = 2, seed = 123) %>%
  map_simulate_occurrences() %>%
  map_sample_observations() %>%
  map_filter_observations() %>%
  map_add_coordinate_uncertainty() %>%
  map_grid_designation(nested = FALSE)  %>%
  select(-all_of(names(multi_species_dataset))) %>%
  select(-occurrences, -observations_total, -observations)
#> [1] [using unconditional Gaussian simulation]
#> [2] [using unconditional Gaussian simulation]
#> [3] [using unconditional Gaussian simulation]
#> [4] [using unconditional Gaussian simulation]
#> [5] [using unconditional Gaussian simulation]
#> [6] [using unconditional Gaussian simulation]
``` 

``` r
# Process cube with b3gbi
gcube_data <- process_cube(
  cube_name = map_occ_cube_df,
  grid_type = "none",
  cols_year = "time_point",
  cols_cellCode = "id",
  cols_occurrences = "n",
  cols_scientificName = "species",
  cols_minCoordinateUncertaintyInMeters = "min_coord_uncertainty",
  cols_kingdom = "kingdom",
  cols_family = "family",
  cols_speciesKey = "species_key"
  
)
gcube_data
#> 
#> Simulated data cube for calculating biodiversity indicators
#> 
#> Date Range: 1 - 5 
#> Number of cells: 
#> Grid reference system: none 
#> Coordinate range:
#> NULL
#> 
#> Total number of observations: 1382 
#> Number of species represented: 6 
#> Number of families represented:  
#> 
#> Kingdoms represented:  
#> 
#> First 10 rows of data (use n = to show more):
#> 
#> # A tibble: 12,000 × 13
#>    scientificName taxonKey genus  family  order class phylum kingdom  year id   
#>    <chr>             <dbl> <chr>  <chr>   <chr> <chr> <chr>  <chr>   <dbl> <chr>
#>  1 species1              1 genus3 family1 orde… clas… phylu… kingdo…     1 106  
#>  2 species1              1 genus3 family1 orde… clas… phylu… kingdo…     1 109  
#>  3 species1              1 genus3 family1 orde… clas… phylu… kingdo…     1 113  
#>  4 species1              1 genus3 family1 orde… clas… phylu… kingdo…     1 117  
#>  5 species1              1 genus3 family1 orde… clas… phylu… kingdo…     1 119  
#>  6 species1              1 genus3 family1 orde… clas… phylu… kingdo…     1 124  
#>  7 species1              1 genus3 family1 orde… clas… phylu… kingdo…     1 131  
#>  8 species1              1 genus3 family1 orde… clas… phylu… kingdo…     1 134  
#>  9 species1              1 genus3 family1 orde… clas… phylu… kingdo…     1 147  
#> 10 species1              1 genus3 family1 orde… clas… phylu… kingdo…     1 154  
#> # ℹ 11,990 more rows
#> # ℹ 3 more variables: obs <dbl>, minCoordinateUncertaintyInMeters <dbl>,
#> #   geometry <POLYGON>
``` 

``` r
# Try calculate time series indicator
total_occ_ts(gcube_data)
#> Biodiversity indicator time series
#> 
#> Name of indicator: Total Occurrences 
#> 
#> Date Range: 1 - 5 
#> 
#> Coordinate range represented:
#> xmin xmax ymin ymax 
#> "NA" "NA" "NA" "NA" 
#> 
#> Number of species represented: 6 
#> Kingdoms represented: NA 
#> 
#> First 10 rows of data (use n = to show more):
#> 
#> # A tibble: 5 × 2
#>    year diversity_val
#>   <dbl>         <dbl>
#> 1     1           244
#> 2     2           326
#> 3     3           198
#> 4     4           282
#> 5     5           332
```

<sup>Created on 2024-07-26 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

shawndove commented:
Apologies again for the long wait! I am back from holidays now.
All the issues mentioned are now solved using the code you provided.

wlangera commented:
Thanks!

Some details:

- with `process_cube()`, it say in the metadata output: `Number of cells: No cell codes provided`. Yet `cols_cellCode = "id"` is provided. So this claim is false and the number of cells is easily calculated with `unique()`
- with `process_cube()`, it say in the metadata output: `Kingdoms represented: kingdom1`, but with `total_occ_ts()`, it says `Kingdoms represented: NA`. So there is a mismatch there

shawndove commented:
Oops, sorry, I forgot to mention that you have to select "custom" instead of "none" now! As for the other issue, should now show kingdoms and number of families for the ts indicator

wlangera reviewed (approved):
Thanks! :)

wlangera commented:
@shawndove if you want, you can give me write access to the repo. This way I can approve this PR, and we can work together more efficiently on this package if needed.

shawndove commented:
Ok, write access granted. Sorry, I'm still learning some of these github features. I didn't know you weren't able to approve the pull request!
